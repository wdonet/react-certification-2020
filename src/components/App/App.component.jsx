import React, { useState } from 'react';
import styled, { ThemeProvider } from 'styled-components';
import { LayoutWrapper } from '../Layout';
import HomeVideos from '../HomeVideos/HomeVideos';
import { lightTheme, darkTheme } from '../../providers/themes';

const StyledWelcome = styled.div`
  text-align: center;
`;

function App() {
  const [isLightTheme, setIsLightTheme] = useState(true);

  const switchTheme = () => {
    setIsLightTheme(!isLightTheme);
  };

  const getThemeConfig = () => ({
    theme: isLightTheme ? lightTheme : darkTheme,
    switchTheme,
  });

  return (
    <ThemeProvider theme={getThemeConfig()}>
      <LayoutWrapper>
        <StyledWelcome>
          <h1>Welcome to the challenge!</h1>
        </StyledWelcome>
        <HomeVideos />
      </LayoutWrapper>
    </ThemeProvider>
  );
}

export default App;
