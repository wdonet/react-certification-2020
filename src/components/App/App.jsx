import React from 'react';
import NavBar from '../NavBar';
import HomePage from '../../pages/Home';

function App() {
  return (
    <div>
      <NavBar />
      <HomePage />
    </div>
  );
}

export default App;
