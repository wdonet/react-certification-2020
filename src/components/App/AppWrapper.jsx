import React, { useReducer } from 'react';
import { appReducer } from '../../reducers/appReducer';
import { lightTheme, darkTheme } from '../../providers/themes';
import youtubeSearchService from '../../services/youtubeSearchService';
import AppContext from '../../providers/AppContext';
import { 
  SET_IS_FIRST_LOAD, 
  SET_SESSION, 
  SET_VIDEOS_LIST, 
  SWITCH_THEME 
} from '../../reducers/actionTypes';

const USER_SESSION_KEY = "user_session";
const initialState = {
  videosList: [],
  currentVideoId: undefined,
  isLightTheme: true,
  isFirstLoad: true,
  userSession: JSON.parse(window.sessionStorage.getItem(USER_SESSION_KEY)),
};

function AppWrapper(props) {
  const [{ 
    videosList, 
    isLightTheme, 
    isFirstLoad,
    userSession}, 
    dispatch] = useReducer(appReducer, initialState);

  const getThemeConfig = () => ({
    theme: isLightTheme ? lightTheme : darkTheme,
    switchTheme: () => dispatch({type: SWITCH_THEME}),
  });

  const search = async (query) => {
    const serviceResponse = await youtubeSearchService.search(query);
    serviceResponse.error 
    ? dispatch({ type: SET_VIDEOS_LIST, payload: [] })
    : dispatch({ type: SET_VIDEOS_LIST, payload: serviceResponse });
  }

  const firstSearch = async () => {
    let counter = 0;
    /* global gapi */
    /* eslint no-undef: "error" */
    while((!gapi || !gapi.client) && counter < 10){
      counter+=1;
      await new Promise(resolve => setTimeout(resolve, 100));
    }
    ((!gapi || !gapi.client) || counter >= 10) 
    ? dispatch({ type: SET_VIDEOS_LIST, payload: [] })
    : search()
  }

  if(isFirstLoad){
    dispatch({ type: SET_IS_FIRST_LOAD, payload: false });
    firstSearch();
  }

  return (
    <AppContext.Provider value={
      { ...getThemeConfig(), 
        search,
        videosList,
        userSession,
        setUserSession: (data) => dispatch({type: SET_SESSION, payload: data}),
      }
    }>
      { props.children }
    </AppContext.Provider>
  );
}

export default AppWrapper;
