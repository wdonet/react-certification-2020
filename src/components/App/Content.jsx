import React from 'react';
import { Redirect, Route, Switch } from 'react-router';
import HomePlayer from '../HomeVideos/HomePlayer';
import HomeVideos from '../HomeVideos/HomeVideos';

const Content = () => {
    return (<div>
        <Switch>
            <Route exact path="/">
                <Redirect to="/home"/>
            </Route>
            <Route path="/home">
                <HomeVideos />
            </Route>
            <Route path="/player">
                <HomePlayer />
            </Route>
        </Switch>
    </div>);
}

export default Content;