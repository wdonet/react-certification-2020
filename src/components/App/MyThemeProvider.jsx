import React from 'react';
import { ThemeProvider } from 'styled-components';
import { useYouTube } from '../YouTube/YouTubeProvider';

const theme = {
  dark: {
    primary: '#4396f3',
    secondary: '#4a4646',
    color: '#f7f7f7',
  },
};

function MyThemeProvider({ children }) {
  const { state } = useYouTube();
  const { theme: stateTheme } = state;
  return <ThemeProvider theme={theme[stateTheme] || {}}>{children}</ThemeProvider>;
}

export default MyThemeProvider;
