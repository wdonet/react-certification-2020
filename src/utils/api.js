export const getVideos = async () => {
  return window.gapi.client.request({
    path: 'https://www.googleapis.com/youtube/v3/videos',
    params: { chart: 'mostPopular', part: 'snippet,statistics', maxResults: 25 },
  });
};

export const searchVideos = async (text) => {
  return window.gapi.client.request({
    path: 'https://www.googleapis.com/youtube/v3/search',
    params: {
      q: text,
      chart: 'mostPopular',
      part: 'snippet',
      maxResults: 25,
      type: 'video',
    },
  });
};

export const getRelatedVideos = async (videoId) => {
  return window.gapi.client.request({
    path: 'https://www.googleapis.com/youtube/v3/search',
    params: {
      part: 'snippet',
      type: 'video',
      maxResults: 25,
      relatedToVideoId: videoId,
    },
  });
};

export const getVideoStatistics = async (videoId) => {
  return window.gapi.client.request({
    path: 'https://www.googleapis.com/youtube/v3/videos',
    params: {
      part: 'snippet,statistics',
      id: videoId,
    },
  });
};
