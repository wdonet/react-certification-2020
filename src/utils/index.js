export { default as contextWrapper } from './testing/helpers/contextWrapper';
export { default as renderWithTheme } from './testing/helpers/renderWithTheme';
export { default as googleMockedAPIObject } from './testing/mocks/googleMockedAPIObject';
export { default as youtubeMockedData } from './testing/mocks/youtubeMockedData';
export { default as YTMockedObject } from './testing/mocks/YTMockedObject';
