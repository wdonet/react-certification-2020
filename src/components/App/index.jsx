import React, { useEffect, useState }from 'react';
import HeaderBar from '../HeaderBar';
import Content from '../Content';
import { items } from '../../assets/mockdata/mockdata.json';
import VideoCard from '../VideoCard';
import filterByYear from '../../utils/filter';

function App() {
  const [recent, setRecent] = useState([]);
  const [inDetail, setInDetail] = useState(false);
  const [detailVideoId, setDetailVideoId] = useState("");
  const [detailTitle, setDetailTitle] = useState("");
  const [detailDescription, setDetailDescription] = useState("");
  
  
  useEffect(() => { 
      console.log('useEffect');
      fetch(`https://www.googleapis.com/youtube/v3/search?key=${process.env.REACT_APP_YOUTUBE_API_KEY2}&q=wizeline&part=snippet&maxResults=50&order=date&type=video`)
      .then((res) => res.json())
      .then((res) => {
        if (res.items) {
          setRecent(res.items);
        } else {
          setRecent(items);
        }
        
      }
    ).catch((emAll) => {});
  }, []);
  
  const UpdateQuery = (text) => {
    console.log(`New query: ${text}`);
    fetch(`https://www.googleapis.com/youtube/v3/search?key=${process.env.REACT_APP_YOUTUBE_API_KEY2}&q=${text}&part=snippet&maxResults=50&order=date&type=video`)
      .then((res) => res.json())
      .then((res) => {
        if (res.items) {
          setRecent(res.items);
        } else {
          setRecent(items);
        }
        setInDetail(false);
      });
  };

  const changeDetail = (indetail=false, videoId, title, description) => {
    // setInDetail(inDetail ? false : true);
    console.log(`title: ${title}`);
    console.log(`description: ${description}`);
    setInDetail(indetail);
    setDetailVideoId(videoId);
    setDetailTitle(title);
    setDetailDescription(description);
  };


  return (
    <React.StrictMode>
      <HeaderBar updateQuery={ UpdateQuery } />
      <Content inDetail={inDetail} detailVideoId={detailVideoId} gotodetail={changeDetail} detailTitle={detailTitle} detailDescription={ detailDescription}>{recent ?
        recent.map((video) => {
          return (
            <VideoCard
              key={video.etag}
              videoId ={video.id.videoId}
              title={video.snippet.title}
              description={video.snippet.description}
              image={video.snippet.thumbnails.high.url}
              gotodetail={ changeDetail }
            />
          )
        })
      :null}
      </Content>
    </React.StrictMode>
  );
}

export default App;
