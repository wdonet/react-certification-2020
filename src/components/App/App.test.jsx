import React from 'react';
import { act, getByRole, render } from '@testing-library/react';
import { googleMockedAPIObject } from '../../utils';
import App from './index';

global.gapi = googleMockedAPIObject();

const build = async (Component = <App />) => {
  let container;
  await act(async () => {
    container = render(Component).container;
  });
  return {
    container,
    LayoutWrapper: () => getByRole(container, 'application'),
    ThemeSwitch: () => getByRole(container, 'switch'),
  };
};

describe('App layout', () => {
  it('renders', async () => {
    const wrapper = await build();
    expect(wrapper).toMatchSnapshot();
  });

  it('displays layout wrapper', async () => {
    const wrapper = await build();
    expect(wrapper.LayoutWrapper()).toBeInTheDocument();
  });
});
