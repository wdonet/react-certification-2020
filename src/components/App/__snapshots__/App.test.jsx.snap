// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App layout renders 1`] = `
Object {
  "LayoutWrapper": [Function],
  "ThemeSwitch": [Function],
  "container": <div>
    <div
      class="sc-VigVT hxPfJy"
      data-testid="layout-wrapper"
      role="application"
    >
      <div
        class="sc-iwsKbI fwoerb"
        data-testid="header"
        role="toolbar"
      >
        <div
          class="sc-gZMcBi hhbgDr"
        >
          <div>
            <img
              alt="hamburguer"
              class="sc-gzVnrw iCjpdg"
              role="button"
              src="hamburguer.png"
            />
          </div>
          <div
            data-testid="search-input"
          >
            <input
              class="sc-dnqmqq eJdbOI"
              role="search"
            />
          </div>
        </div>
        <div
          class="sc-gZMcBi hhbgDr"
        >
          <div
            data-testid="theme-mode-switch"
          >
            <input
              class="sc-htoDjs bZzhDu"
              role="switch"
              type="checkbox"
            />
          </div>
          <div
            data-testid="user-avatar"
          >
            <img
              alt="profile"
              class="sc-bdVaJa ieaWl"
              role="button"
              src="default_user.jpg"
            />
          </div>
        </div>
      </div>
      <div
        class="sc-gqjmRU bMokHB"
        data-testid="main-content"
        role="main"
      >
        <div
          class="sc-fjdhpX fJdZy"
        >
          <h1>
            Welcome to the challenge!
          </h1>
        </div>
        <div
          class="sc-jTzLTM dnPpCj"
        >
          <div
            data-testid="video-card-{\\"kind\\":\\"youtube#channel\\",\\"channelId\\":\\"UCPGzT4wecuWM0BH9mPiulXg\\"}"
          >
            <div
              class="sc-bwzfXH fFIYHG"
              role=""
            >
              <figure
                class="sc-ifAKCX kHuOca"
              >
                <img
                  alt="Wizeline"
                  class="sc-EHOje hZoKKp"
                  src="https://yt3.ggpht.com/ytc/AAUvwnighSReQlmHl_S_vSfvnWBAG5Cw4A0YxtE0tm5OpQ=s800-c-k-c0xffffffff-no-rj-mo"
                />
                <figcaption
                  hidden=""
                >
                  Wizeline transforms how teams build technology. Its customers accelerate the delivery of innovative products with proven solutions, which combine Wizeline's ...
                </figcaption>
              </figure>
              <div
                class="sc-htpNat jRJAUB"
              >
                <h2
                  class="sc-bZQynM iihQAR"
                  title="video-title"
                >
                  Wizeline
                </h2>
                <p
                  class="sc-bxivhb iKRSeQ"
                  role=""
                >
                  Wizeline transforms how teams build technology. Its customers accelerate the delivery of innovative products with proven solutions, which combine Wizeline's ...
                </p>
              </div>
            </div>
          </div>
          <div
            data-testid="video-card-{\\"kind\\":\\"youtube#video\\",\\"videoId\\":\\"nmXMgqjQzls\\"}"
          >
            <div
              class="sc-bwzfXH fFIYHG"
              role=""
            >
              <figure
                class="sc-ifAKCX kHuOca"
              >
                <img
                  alt="Video Tour | Welcome to Wizeline Guadalajara"
                  class="sc-EHOje hZoKKp"
                  src="https://i.ytimg.com/vi/nmXMgqjQzls/hqdefault.jpg"
                />
                <figcaption
                  hidden=""
                >
                  Follow Hector Padilla, Wizeline Director of Engineering, for a lively tour of our office. In 2018, Wizeline opened its stunning new office in Guadalajara, Jalisco, ...
                </figcaption>
              </figure>
              <div
                class="sc-htpNat jRJAUB"
              >
                <h2
                  class="sc-bZQynM iihQAR"
                  title="video-title"
                >
                  Video Tour | Welcome to Wizeline Guadalajara
                </h2>
                <p
                  class="sc-bxivhb iKRSeQ"
                  role=""
                >
                  Follow Hector Padilla, Wizeline Director of Engineering, for a lively tour of our office. In 2018, Wizeline opened its stunning new office in Guadalajara, Jalisco, ...
                </p>
              </div>
            </div>
          </div>
          <div
            data-testid="video-card-{\\"kind\\":\\"youtube#video\\",\\"videoId\\":\\"HYyRZiwBWc8\\"}"
          >
            <div
              class="sc-bwzfXH fFIYHG"
              role=""
            >
              <figure
                class="sc-ifAKCX kHuOca"
              >
                <img
                  alt="Wizeline Guadalajara | Bringing Silicon Valley to Mexico"
                  class="sc-EHOje hZoKKp"
                  src="https://i.ytimg.com/vi/HYyRZiwBWc8/hqdefault.jpg"
                />
                <figcaption
                  hidden=""
                >
                  Wizeline continues to offer a Silicon Valley culture in burgeoning innovation hubs like Mexico and Vietnam. In 2018, our Guadalajara team moved into a ...
                </figcaption>
              </figure>
              <div
                class="sc-htpNat jRJAUB"
              >
                <h2
                  class="sc-bZQynM iihQAR"
                  title="video-title"
                >
                  Wizeline Guadalajara | Bringing Silicon Valley to Mexico
                </h2>
                <p
                  class="sc-bxivhb iKRSeQ"
                  role=""
                >
                  Wizeline continues to offer a Silicon Valley culture in burgeoning innovation hubs like Mexico and Vietnam. In 2018, our Guadalajara team moved into a ...
                </p>
              </div>
            </div>
          </div>
          <div
            data-testid="video-card-{\\"kind\\":\\"youtube#video\\",\\"videoId\\":\\"Po3VwR_NNGk\\"}"
          >
            <div
              class="sc-bwzfXH fFIYHG"
              role=""
            >
              <figure
                class="sc-ifAKCX kHuOca"
              >
                <img
                  alt="Wizeline hace sentir a empleados como en casa"
                  class="sc-EHOje hZoKKp"
                  src="https://i.ytimg.com/vi/Po3VwR_NNGk/hqdefault.jpg"
                />
                <figcaption
                  hidden=""
                >
                  En el 2014, Bismarck fundó Wizeline, compañía tecnológica que trabaja con los corporativos ofreciendo una plataforma para que desarrollen software de forma ...
                </figcaption>
              </figure>
              <div
                class="sc-htpNat jRJAUB"
              >
                <h2
                  class="sc-bZQynM iihQAR"
                  title="video-title"
                >
                  Wizeline hace sentir a empleados como en casa
                </h2>
                <p
                  class="sc-bxivhb iKRSeQ"
                  role=""
                >
                  En el 2014, Bismarck fundó Wizeline, compañía tecnológica que trabaja con los corporativos ofreciendo una plataforma para que desarrollen software de forma ...
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
}
`;
